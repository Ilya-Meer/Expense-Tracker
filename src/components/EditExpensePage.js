import React, { Component } from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { H2, Button } from './StyledComponents/Shared';
import { startEditExpense, startRemoveExpense } from '../actions/expenseActions';

class EditExpensePage extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
    this.onClick = this.onClick.bind(this);
  }
  onSubmit(expense) {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  onClick(expense) {
    this.props.startRemoveExpense(this.props.expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <H2>Edit Expense</H2>
        <ExpenseForm
          expense={this.props.expense} 
          onSubmit={this.onSubmit}
        />
        <Button
        onClick={this.onClick}
        negative={true}
        style={{margin: '10px 5%'}}
        >
        Remove Expense
      </Button>
      </div>
    )
  }
};

export {EditExpensePage}


const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id) 
  }
};

const mapDispatchToProps = (dispatch) => {
  return {
    startEditExpense: (id, expense) => {
      dispatch(startEditExpense(id, expense));
    }, 
    startRemoveExpense: (expense) => {
      dispatch(startRemoveExpense(expense))
    }
  }
};


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);