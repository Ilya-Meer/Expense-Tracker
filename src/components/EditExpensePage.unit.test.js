import React from 'react';
import { shallow } from 'enzyme';
import {EditExpensePage} from './EditExpensePage';
import moment from './__mocks__/moment'

describe('Edit Expense Page', () => {
  let onSubmit, onClick, history, wrapper, expense;
  beforeEach(() => {
    onSubmit = jest.fn();
    onClick = jest.fn();
    history = {
      push: jest.fn()
    }
    expense = {
      id: '3',
      desc: 'Rent', 
      note: 'The rent was not cheap either', 
      amount: '2000000', 
      createdAt: moment(0)
    };
    wrapper = shallow(
      <EditExpensePage 
        expense={expense} 
        startEditExpense={onSubmit} 
        startRemoveExpense={onClick} 
        history={history}
      />
    );
  })

  it('should render properly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should handle edit Expense', () => {
     wrapper.find('ExpenseForm').prop('onSubmit')(expense);
     expect(onSubmit).toHaveBeenLastCalledWith(expense.id, expense);
     expect(history.push).toHaveBeenLastCalledWith('/');
  })

  it('should handle remove Expense', () => {
     wrapper.find('button').prop('onClick')(expense);
     expect(onClick).toHaveBeenLastCalledWith(expense);
     expect(history.push).toHaveBeenLastCalledWith('/');
  })
})