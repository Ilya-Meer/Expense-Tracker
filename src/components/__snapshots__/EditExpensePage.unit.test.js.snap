// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Expense Page should render properly 1`] = `
<div>
  <styled.h2>
    Edit Expense
  </styled.h2>
  <ExpenseForm
    expense={
      Object {
        "amount": "2000000",
        "createdAt": "1970-01-01T00:00:00.000Z",
        "desc": "Rent",
        "id": "3",
        "note": "The rent was not cheap either",
      }
    }
    onSubmit={[Function]}
  />
  <styled.button
    negative={true}
    onClick={[Function]}
    style={
      Object {
        "margin": "10px 5%",
      }
    }
  >
    Remove Expense
  </styled.button>
</div>
`;
